### ISSUE TITLE
AWS4Signer produces incorrect signature for reserved characters

### ISSUE THREAD

**Issue by jarruda** [aws/aws-sdk-net#1943]
## Description
[AWS4Signer](https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4Signer.cs#L30) produces an incorrect signature for URIs containing reserved characters.

The root cause is [AWSSDKUtils.CanonicalizeResourcePath](https://github.com/aws/aws-sdk-net/blob/db5c9aec40d2d1f9b8f97d34e95a9b8108c39558/sdk/src/Core/Amazon.Util/AWSSDKUtils.cs#L331) incorrectly encoding path components when `detectPreEncode` is true -- which `AWS4Signer` sets for all non-S3 URIs.

I'm aware `AWS4Signer` is an internal class, but unlike other AWS SDKs it doesn't appear that a publicly-vended SigV4 signer is part of the .NET SDK, so I have attempted to use it.  Nonetheless, the root cause is a utility function in the public portion of the SDK.  Please correct me if this is not the case and there is a publicly vended signing utility.

## Reproduction Steps

Prepare a request containing a reserved character such as `:` and attempt to have that request signed by AWS4Signer.  An easier way to isolate the issue is to directly use `AWSSDKUtils.CanonicalizeResourcePath` in the same way it does.  That is, `detectPreEncode` must be set to `true`.

For an example path `/:`, or `/%3A` once encoded to a legal URI:

```cs
var encodedPathComponent = AWSSDKUtils.UrlEncode(":", false);
var path = $"/{encodedPathComponent}";
var canonicalPath = AWSSDKUtils.CanonicalizeResourcePath(new Uri("https://execute-api.us-east-1.amazonaws.com"), path, true);

// path = "/%3A"
// canonicalPath = "/%25253A" -- should be "%253A"     
```

One could assume that the output is being encoded twice per the [SigV4 signing protocol](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html), and that the input should simply be given in an unencoded form and let the canonicalization process handle it.  Unfortunately, this does not work because the [UrlEncode](https://github.com/aws/aws-sdk-net/blob/db5c9aec40d2d1f9b8f97d34e95a9b8108c39558/sdk/src/Core/Amazon.Util/AWSSDKUtils.cs#L363) pass that is performed as a result of `detectPreEncode` being set does not encode reserved characters because the `path` argument is set to `true` in that call.  Passing an unencoded path then results in reserved characters not being encoded in the first pass, only in the second pass, and the canonical resource path not being correct either:

```cs
var path = $"/:";
var canonicalPath = AWSSDKUtils.CanonicalizeResourcePath(new Uri("https://execute-api.us-east-1.amazonaws.com"), path, true);

// path = "/:" -- must be "/%3A" for the actual request to be a valid URI
// canonicalPath = "%3A" -- should be "%253A"
```

## Resolution
- [ ] :wave: I can/would-like-to implement a fix for this problem myself

---

This is a :bug: bug-report

**Comment by peterrsongg** [comment]
@jarruda We recently fixed an encoding issue in this release on 10-27 [Core 3.7.204.10](https://github.com/aws/aws-sdk-net-staging/blob/main/changelogs/SDK.CHANGELOG.2023.md#376720-2023-10-27-1814-utc) . Do you mind testing your scenario with that version of core?

**Comment by jarruda** [comment]
@peterrsongg -- I no longer have a C# environment set up I can test in.  The repro test is only 3 lines from the description if you would like to see if it still reproduces.

**Comment by peterrsongg** [comment]
@jarruda  The latest version of the SDK uses CanonicalizeResourcePathV2 and that is where we made the encoding fix. 

I used the repro test from above and confirmed that the latest version of the SDK fixes this issue. It seems in your repro code you passed in an already encoded `:` which you then pass in again to `CanonicalizeResourcePath`. In the latest version that is not necessary. The SDK will encoded special characters by default so all you have to do is:

```
var path = $"/:";
var canonicalPath = AWSSDKUtils.CanonicalizeResourcePathV2(new Uri("https://execute-api.us-east-1.amazonaws.com"), path, true, null);
Console.WriteLine(canonicalPath);
```
which will return the desired `/%253A`

The second code snippet also works as  expected:
```
var path = $"/:";
var canonicalPath = AWSSDKUtils.CanonicalizeResourcePathV2(new Uri("https://execute-api.us-east-1.amazonaws.com"), path, true, null);
Console.WriteLine(canonicalPath);
```
which returns the desired result : `/%253A`. Thank you for bringing this to our attention

**Comment by jarruda** [comment]
Note this issue is specifically about the behavior of `AWS4Signer` whose use of the CanonicalizeResourcePath API resulted in the double-encoding.  My repro focused on what I identified as the root cause of the incorrect encoding inside of this class.

So the question is, does AWS4Signer correctly sign requests which include the `:` character in the resource path (when an S3 host is not used)?  Sorry I do not have a simple repro for you to follow; you would have to calculate the correct signature for the request and compare that to the Signer's output to check.  This being incorrect is how I diagnosed the initial problem.

**Comment by peterrsongg** [comment]
@jarruda Yes, the AWS4Signer does correctly sign requests which include special characters, we had to create V2 versions of the helper methods that CanonicalizeResourcePathV2 uses (which bubbles up to AWS4Signer). We do double encode these special characters. [This changelog entry](https://github.com/aws/aws-sdk-net/blob/main/changelogs/SDK.CHANGELOG.2023.md#376720-2023-10-27-1814-utc) is where the fix was introduced. As part of the implementation for that we tested quite a bit with the signer
